### PR TITLE
Add useUrlState to change both Route Params and Query String

### LINKED ISSUE #78: Setting both query string key and route params in one render/effect does not work.
Steps to reproduce
1. Run the following app 
2. Go to `/test/old-slug`
3. Observe the param is added but slug is still old: /test/old-slug?param=new-param  
4. If you reverse the order of setSlugs/setParam, the opposite happens - slug is updated but param not.   

```
import {useEffect} from "react";
import {Route, BrowserRouter, Switch} from "react-router-dom";
import {useQueryStringKey, useRouteParams} from "use-route-as-state";

function TestSlug() {
    const [slugs, setSlugs] = useRouteParams();
    const [param, setParam] = useQueryStringKey('param')

    useEffect(() => {
        setSlugs({slug: 'new-slug'});
        setParam('new-param');
    }, [])

    return <span> Slug: {slugs.slug} Param: {param}</span>
}

function App() {
    return (
            <BrowserRouter>
                <Switch>
                    <Route path="/test/:slug"><TestSlug /></Route>
                </Switch>
            </BrowserRouter>
    )
}
```

They affect the same underlying state (URL string) so this might be actually a result how React works.
A great workaround would be to provide 'useUrlState' allowing to set both route params and query string keys through single setter.

### PR DESCRIPTION
**Pull request by baruchiro**
Fixes #78

### CODE REVIEW

**Comment by baruchiro** [comment]
@alex-kowalczyk, @itayganor Please review

File: README.md [hunk 1]
```diff
@@ -45,12 +45,13 @@ const Example = () => {
 This library is trying to behave like the `useState` React hook, by exposing a similar interface.
 
 ```typescript
-type DisaptchState = Dispatch<SetStateAction<Record<string, string>>>
+type DisaptchState<TState> = Dispatch<SetStateAction<TState>>
```
Comment (itayganor) [inline README.md:48]: `DisaptchState` => `DispatchState`

File: README.md [hunk 2]
```diff
@@ -106,6 +107,21 @@ const Example = () => {
...
+```typescript
+type UrlState = {
+  params: RouteObject,
+  query: RouteObject
+}
+```
+
+> **Type:** `useUrlState: (defaultValues?: UrlState): [UrlState, DisaptchState<UrlState>]`
+
+Due to limitation in *React Router*, and *React* itself, you can't use different hooks here together during one render cycle.
```
Comment (itayganor) [inline README.md:121]:
```suggestion
Due to limitations in *React Router*, and *React* itself, you can't use different hooks here together during one render cycle.
```

File: README.md [hunk 3]
```diff
@@ -106,6 +107,21 @@ const Example = () => {
...
+
+In order to solve that, you can use this hook to control both *route params* and *qeury string* at once.
```
Comment (itayganor) [inline README.md:123]:
```suggestion
In order to solve that, you can use this hook to control both *route params* and *query string* at once.
```

File: src/index.ts [hunk 4]
```diff
@@ -4,13 +4,14 @@ import { encodeValues, useDecodedLocation, useDecodedRouteMatch } from './encode
 import { objectToQueryParams, removeUndefined } from './helpers'
 export * from './deprecated'
 
-type DisaptchState = Dispatch<SetStateAction<Record<string, string>>>
+type DisaptchState<TState> = Dispatch<SetStateAction<TState>>
```
Comment (itayganor) [inline src/index.ts:7]: `DisaptchState` => `DispatchState`

File: README.md (hunk 1 above)
Comment (baruchiro) [reply README.md:48]:
```suggestion
type DispatchState<TState> = Dispatch<SetStateAction<TState>>
```

File: src/index.ts (hunk 4 above)
Comment (baruchiro) [reply src/index.ts:7]:
```suggestion
type DispatchState<TState> = Dispatch<SetStateAction<TState>>
```

**Pull request merged**